packages/grid/x-data-grid/src/components/GridRow.tsx

```diff
@@ -292,110 +400,51 @@ function GridRow(props: React.HTMLAttributes<HTMLDivElement> & GridRowProps) {
       ? rootProps.showCellRightBorder
       : !removeLastBorderRight && rootProps.disableExtendRowFullWidth;
 
-    const cellParams = apiRef.current.getCellParams(rowId, column.field);
-
-    const classNames: string[] = [];
-
-    const disableDragEvents =
-      (rootProps.disableColumnReorder && column.disableReorder) ||
-      (!(rootProps as any).rowReordering &&
-        !!sortModel.length &&
-        treeDepth > 1 &&
-        Object.keys(editRowsState).length > 0);
-
-    if (column.cellClassName) {
-      classNames.push(
-        clsx(
-          typeof column.cellClassName === 'function'
-            ? column.cellClassName(cellParams)
-            : column.cellClassName,
-        ),
+    if (row) {
+      const cellColSpanInfo = apiRef.current.unstable_getCellColSpanInfo(
+        rowId,
+        indexRelativeToAllColumns,
       );
-    }
 
-    const editCellState = editRowsState[rowId] ? editRowsState[rowId][column.field] : null;
-    let content: React.ReactNode = null;
-
-    if (editCellState == null && column.renderCell) {
-      content = column.renderCell({ ...cellParams, api: apiRef.current });
-      // TODO move to GridCell
-      classNames.push(
-        clsx(gridClasses['cell--withRenderer'], rootProps.classes?.['cell--withRenderer']),
+      if (cellColSpanInfo && !cellColSpanInfo.spannedByColSpan) {
+        const { colSpan, width } = cellColSpanInfo.cellProps;
+        const cellProps = { width, colSpan, showRightBorder, indexRelativeToAllColumns };
+        cells.push(getCell(column, cellProps));
+      }
+    } else {
+      const cellColSpanInfo = apiRef.current.unstable_getCellColSpanInfo(
+        rowId,
+        indexRelativeToAllColumns,
       );
```

Comment:
`unstable_getCellColSpanInfo` call can be deduplicated and moved up outside of `if (row) {` statement